### PR TITLE
Update to mfem@4.5

### LINKED ISSUE #974: Update to mfem@4.5
MFEM modified its `QuadratureSpace` and `QuadratureFunction` API between v4.4 and v4.5.

We use these in our `MFEMSidreDataCollection` class and in quest's sample-based shaping application, and users ran into compilation errors when using axom with `mfem@4.5`

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR is a feature
- It updates our MFEM TPL from `v4.4` to `v4.5` to resolve API changes to mfem's `QuadratureFunction` and `QuadratureSpace` classes
  - Resolves #974 

#### TODO
- [x] Update spack config
- [x] Update RZ host-configs
- [x] Update CZ host-configs
- [x] Update docker images
- [x] Update windows/vcpkg build
- [x] Update RELEASE-NOTES

### CODE REVIEW

File: host-configs/ruby-toss_3_x86_64_ib-clang@10.0.0.cmake [hunk 1]
```diff
@@ -69,17 +69,17 @@ set(ENABLE_GTEST_DEATH_TESTS ON CACHE BOOL "")
...
 
 set(C2C_DIR "${TPL_ROOT}/c2c-1.3.0" CACHE PATH "")
 
-set(MFEM_DIR "${TPL_ROOT}/mfem-4.4.0" CACHE PATH "")
+set(MFEM_DIR "${TPL_ROOT}/mfem-4.5.0" CACHE PATH "")
 
 set(HDF5_DIR "${TPL_ROOT}/hdf5-1.8.22" CACHE PATH "")
 
-set(LUA_DIR "${TPL_ROOT}/lua-5.3.5" CACHE PATH "")
+set(LUA_DIR "/usr/WS1/axom/libs/toss_3_x86_64_ib/2022_12_08_21_44_21/clang-9.0.0/lua-5.4.4" CACHE PATH "")
```
Comment (kennyweiss) [inline host-configs/ruby-toss_3_x86_64_ib-clang@10.0.0.cmake:82]: Note: These sorts of inconsistencies (e.g. clang@9 in a clang@10 host-config) should be fixed when we update uberenv

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp [hunk 2]
```diff
@@ -2644,12 +2640,18 @@ void MFEMSidreDataCollection::reconstructField(Group* field_grp)
...
-                                     vdim));
+        new mfem::QuadratureFunction(m_quadspaces.at(basis_name).get(), vdim));
+
+      // FIXME: A bug was fixed in QuadratureFunction::SetSpace after mfem@4.5
+      // See: https://github.com/mfem/mfem/pull/3281
+      // Update using the following when possible to avoid an unnecessary allocation in the constructor
+      // (1) Use default Quadrature function constructor
+      // (2) Use qf->SetSpace(qs, values, vdim);
+      auto* qf = m_owned_quadfuncs.back().get();
+      qf->NewDataAndSize(values, vdim * qf->GetSpace()->GetSize());
```
Comment (kennyweiss) [inline src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: Note: mfem refactored the `QuadratureFunction` API between mfem@4.4 and mfem@4.5, removing the constructor that allowed directly passing in the data. Presumably, this was not intentional. I will create a PR in mfem to add it back so we can go back to the simpler form for the next release of mfem.

There was also a bug in the mfem@4.5 release that does not allow us to use the `QuadratureFunction::SetSpace` function. It was fixed in https://github.com/mfem/mfem/pull/3281, but that was merged after the mfem@4.5 release.

As a consequence, it looks like we'll have an unnecessary allocation for the underlying vector in the meantime.

Tag: @tzanio @v-dobrev @rcarson3 @jamiebramwell

**Comment by kennyweiss** [comment]
This branch also passed our Windows TPL build: https://github.com/LLNL/axom/actions/runs/3661815967

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp (hunk 2 above)
Comment (v-dobrev) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: This is unfortunate omission in MFEM v4.5.

If you want to use an MFEM release version with this issue fixed, maybe we can create a bugfix release, v4.5.2, after bringing back the constructor.

Comment (kennyweiss) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: Thanks for the offer @v-dobrev! We'd be happy to upgrade axom's mfem once https://github.com/mfem/mfem/pull/3362 gets merged. 

As far as I can tell, the current workaround will require an unnecessary memory allocation for the data associated with each `QuadratureField` during a restart. I think this affects @jamiebramwell's code more than mine, but I wouldn't expect it to noticeably affect the overall runtime/performance.

**Review by gunney1 — APPROVED** [review]
Thanks @kennyweiss

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp (hunk 2 above)
Comment (jamiebramwell) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: I need to rework our quadrature function storage to work on arbitrary types anyway, so this shouldn't affect us.

**Review by bmhan12 — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp (hunk 2 above)
Comment (v-dobrev) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: @kennyweiss, when you refer to _"axom's mfem"_, how is this mfem version (or commit) set?

Comment (v-dobrev) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: Looking at this PR, it seems that the required version of MFEM is precisely the v4.5 release and that seems to be hard-coded in several places. So with this PR, the only acceptable MFEM version becomes v4.5, right?

Comment (kennyweiss) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: We're currently only maintaining support for a single version of mfem, which we set in our spack config and list in our `RELEASE-NOTES` (see the diffs in this PR).  

If we needed to, I suppose we could add support for different/multiple mfem versions via `#ifdef`s.

Comment (v-dobrev) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:2651]: Okay, I think I understand -- in order to get bugfixes in MFEM available for use in Axom, we need to make them part of a release.

Let us discuss a possible MFEM v4.5.2 release and we'll let you know when it's ready to be tested with Axom before we make the release.

**Pull request merged**